README
Tomboy Reminder Plugin
----------------------

Requirements
------------
This plugin requires:
tomboy-0.10 or higher

Build and Install
-----------------
Simply run

./configure
make
make install (as root)

It should install the tomboy-reminder in the tomboy addins directory.
You can then enable it via the tomboy interface.

Using the Reminder
------------------
To use the plugin simply write a note and when you want to be reminded for
this note, you have two options:
    - put somewhere a "!" followed by a date at which tomboy will
      show a popup with the note content at specified date.
    - write "remind date" or "alert date" or "rappel date"
    
The date format is quite loose you can write anything like:
	4th july
	2005/05/03
	5th
	3 september 05
	next monday
	monday

Starting from v0.5 you can alo use time in the date, both 12- and 24-hour times
are recognized, for example:
	12am
	23:34
	10h
	10h am

You can also combine the two:
	3 july at 12am
	2004/04/02 a 22:50
	monday @ 2h am

In general just write down the date and time as you would naturally it should
work, if not, it's a bug ;)

You can NOT use the "american" or "european" date format "mm/dd/yyyy"
or "dd/mm/yyy" because it is not clear wether the first two digits are
months or days.

Tomboy will scan all the notes shortly after being started and when a note is closed.
* When a date is found, the note will popup that date starting at 00:00 and every
  30min when you close it.
* When a time is found the note will popup at the given time the same day, only once
* When both time and date are found, the note will popup the specified date at the
  specified time, only once

Please submit me by mail your problems/comments/suggestions/bugs/etc (in AUTHORS file)


NEWS
See the ChangeLog !


src/MagicDate.cs
//
// MagicDate.cs: Parses natural langage dates.
//
// Copyright 2005,  Raphaël Slinckx <raphael@slinckx.net>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
// http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Adapted in part from :
// 'Magic' date parsing, by Simon Willison (6th October 2003)
// http://simon.incutio.com/archive/2003/10/06/betterDateInput
//

// $Id: MagicDate.cs,v 1.2 2005/04/11 11:29:05 kikidonk Exp $

using System;
using System.Text;
using System.Text.RegularExpressions;
using System.Collections;

namespace MagicString {
	public class MagicDate {

		private static Hashtable MONTH_NUMBERS = new Hashtable();
		private static string MONTH_PREFIXES = "jan|feb|mar|apr|may|jun|jul|aug|sep|oct|nov|dec|" + //English
									   "jan|fev|mar|avr|mai|jui|jui|aou|sep|oct|nov|dec|" + //French
									   "янв|фев|мар|апр|май|июн|июл|авг|сеп|окт|нов|дек";//Russian
		
		private static Hashtable DAY_NUMBERS = new Hashtable();
		private static string DAY_PREFIXES = "mon|tue|wed|thu|fri|sat|sun|" + //English
			    					 "lun|mar|mer|jeu|ven|sam|dim|" + //French
			    					 "пн|вт|ср|чт|пт|сб|вс"; //Russian
									   							   
		private static string DAYS_PARTICLE = "st|nd|rd|th";
		private static string NEXT = "next|suivant|prochain";
		private static string TODAY = "(today)|(aujourd.?...)";
		
		private static Regex[] GENERIC = {
		   	//4th | 2nd | 3rd | 1st | 10th | 2 | 32
	    	new Regex ("^(?<daynb>\\d{1,2})("+DAYS_PARTICLE+")?$", RegexOptions.IgnoreCase | RegexOptions.Compiled ),
	    	//4th Jan | 4 Jan | 4 january | 4th,january 
	    	new Regex ("^(?<daynb>\\d{1,2})(?:"+DAYS_PARTICLE+")?\\W+(?<monthname>\\w{3,})$", RegexOptions.IgnoreCase | RegexOptions.Compiled),
	    	//4th Jan 2003 | 4 jan 03
	    	new Regex ("^(?<daynb>\\d{1,2})(?:"+DAYS_PARTICLE+")?\\W+(?<monthname>\\w{3,})\\W+(?<yearnb>\\d{2}\\d{2}?)$", RegexOptions.IgnoreCase | RegexOptions.Compiled),
	    	//Jan 4th
	    	new Regex ("^(?<monthname>\\w{3,})\\W+(?<daynb>\\d{1,2})(?:"+DAYS_PARTICLE+")?$", RegexOptions.IgnoreCase | RegexOptions.Compiled),
	    	// Jan 4th 2003
	    	new Regex ("^(?<monthname>\\w{3,})\\W+(?<daynb>\\d{1,2})(?:"+DAYS_PARTICLE+")?\\W+(?<yearnb>\\d{2}\\d{2}?)$", RegexOptions.IgnoreCase | RegexOptions.Compiled),
	    	// yyyy-mm-dd | yyyy/mm/dd | yy mm dd
	    	new Regex ("^(?<yearnb>\\d{2}\\d{2}?)\\W(?<monthnb>\\d{1,2})\\W(?<daynb>\\d{1,2})$", RegexOptions.IgnoreCase | RegexOptions.Compiled)
	    };
	    
	    private static Regex[] RELATIVE = {
	    	// today
	    	new Regex ("^(?<today>"+TODAY+")$", RegexOptions.IgnoreCase | RegexOptions.Compiled),
	    	
	    	// next day_of_week | week | 
	    	new Regex ("^(?<next>"+NEXT+")\\W+(?<day>\\w{3,})$", RegexOptions.IgnoreCase | RegexOptions.Compiled),
	    	new Regex ("^(?<day>\\w{3,})\\W+(?<next>"+NEXT+")$", RegexOptions.IgnoreCase | RegexOptions.Compiled)
	    };
	    
	    public MagicDate () {
	    	//Store month name/index
	    	string[] months = MONTH_PREFIXES.Split('|');
	    	int i = 0;
	    	foreach(string month in months) {
	    		MONTH_NUMBERS[month] = (i++ % 12) + 1;
	    	}
	    	
	    	string[] days = DAY_PREFIXES.Split('|');
	    	i = 0;
	    	foreach(string day in days) {
	    		DAY_NUMBERS[day] = (i++ % 7);
	    	}
		}
		
		public DateTime ParseDate(string date, DateTime baseDate) {
			date = date.Trim();
			
			DateTimeWrapper result;	
			
			//Parse a regular date
			foreach(Regex regex in GENERIC) {
				result = Match(regex, date, baseDate);
				if(result != null)
					return result.Date;
			}
			
			//Parse a reltive date
			foreach(Regex regex in RELATIVE) {
				result = MatchRelative(regex, date, baseDate);
				if(result != null)
					return result.Date;
			}
			
			throw new NoMagicMatchesException("No date found in string '"+date+"'");
		}
		
		public DateTime ParseDate(string date) {
			return ParseDate(date, DateTime.Now);
	    }
	    
	    private DateTimeWrapper MatchRelative(Regex regex, string date, DateTime baseDate) {
	    	Match match = regex.Match (date);
			if(match.Success) {
				int month = baseDate.Month;
				int year = baseDate.Year;
				int day = baseDate.Day;
				
				if (match.Groups["today"].Success) {
					return new DateTimeWrapper (new DateTime (year, month, day));
				}
				
				DayOfWeek dayofweek = baseDate.DayOfWeek;
				int daynb;
				switch(dayofweek) {
					case DayOfWeek.Monday:    daynb = 0; break;
					case DayOfWeek.Tuesday:   daynb = 1; break;
					case DayOfWeek.Wednesday: daynb = 2; break;
					case DayOfWeek.Thursday:  daynb = 3; break;
					case DayOfWeek.Friday:    daynb = 4; break;
					case DayOfWeek.Saturday:  daynb = 5; break;
					case DayOfWeek.Sunday:    daynb = 6; break;
					default: daynb = -1;break;
				}
				
				int parsedNb = ParseDay( match.Groups["day"].ToString() );
				if(parsedNb == -1)
					return null;
					
				int deltaDays = daynb - parsedNb;
				deltaDays = (deltaDays < 0) ? -deltaDays : (6-deltaDays)+1;
					
				return new DateTimeWrapper (new DateTime (year, month, day).AddDays(deltaDays));
			}
			
			return null;
	    }
	    
	    private DateTimeWrapper Match(Regex regex, string date, DateTime baseDate) {
	    	Match match = regex.Match (date);
			if(match.Success) {
				int month = baseDate.Month;
				int year = baseDate.Year;
				
				int day = Int32.Parse (match.Groups["daynb"].ToString());
				
				string monthstring = match.Groups["monthname"].ToString();
				if(monthstring != "") {
					month = ParseMonth(monthstring);
					if(month == -1)
						return null;
				}
				else if((monthstring = match.Groups["monthnb"].ToString()) != "") {
					month = Int32.Parse (monthstring);
				}
				
				string yearstring = match.Groups["yearnb"].ToString();
				if(yearstring != "") {
					year = Int32.Parse (yearstring);
				}
				
				return new DateTimeWrapper (new DateTime (year, month, day));
			}
			
			return null;
		}
		
		private string ReplaceAccents(string s) {
			return s.Replace("é", "e").Replace("ô", "o");
		}
		
		private int ParseMonth(string month) {
			month = ReplaceAccents(month.Substring(0, 3));
			object monthnb = MONTH_NUMBERS[month.ToLower()];
			if(monthnb == null)
				return -1;
			
			return (int) monthnb ;
		}
		
		private int ParseDay(string day) {
			day = day.Substring(0, 3);
			object daynb = DAY_NUMBERS[day.ToLower()];
			if(daynb == null)
				return -1;
			
			return (int) daynb ;
		}
		
		/*
		public static void Main (string[] args)
		{
			MagicDate md = new MagicDate();
			Console.WriteLine(md.ParseDate(args[0]));
		}
		*/
	}
}


src/MagicParser.cs
//
// MagicParser.cs: Convenient methods for MagicDate+Time
//
// Copyright 2005,  Raphaël Slinckx <raphael@slinckx.net>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
// http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

// $Id: MagicParser.cs,v 1.1 2005/04/11 11:29:30 kikidonk Exp $

using System;
using System.Text;
using System.Text.RegularExpressions;
using System.Collections;

namespace MagicString {
	public class MagicParser {
		private MagicDate md = new MagicDate();
		private MagicTime mt = new MagicTime();
		
		private static string DATE_TIME_SEP = "at|a|à|@";
		
		private static Regex[] GENERIC = {
			new Regex("^(?<date>.*)\\W+("+DATE_TIME_SEP+")\\W+(?<time>.*)$", RegexOptions.IgnoreCase | RegexOptions.Compiled),
			new Regex("^(?<date>.*)$", RegexOptions.IgnoreCase | RegexOptions.Compiled),
		};
		
		public ParseResult Parse(string s, DateTime baseDate) {
			s = s.Trim();
			
			ParseResult result;
			
			//Parse a regular date
			foreach(Regex regex in GENERIC) {
				result = Match(regex, s, baseDate);
				if(result != null)
					return result;
			}
			
			throw new NoMagicMatchesException("No matches found in string '"+s+"'");
		}
		
		private ParseResult Match(Regex regex, string s, DateTime baseDate) {
			ParseResult parseResult = null;
			
	    	Match match = regex.Match (s);
			if(match.Success) {
				string date = match.Groups["date"].ToString();
				string time = match.Groups["time"].ToString();
				
				parseResult = new ParseResult ();
				
				DateTime dateTime = baseDate;
				if(date != "" && time != "") {
					dateTime = md.ParseDate (date, baseDate);
					dateTime = mt.AddTime (time, dateTime);
					
					parseResult.HasTime = true;
					parseResult.HasDate = true;
				}
				else if(date != "" && time == "") {
					//either a date only or a time only
					try {
						dateTime = 	md.ParseDate (date, baseDate);
						parseResult.HasDate = true;
					} catch (Exception e) {
						dateTime = mt.AddTime(date, baseDate);
						parseResult.HasTime = true;
					}
				}
				
				parseResult.Date = dateTime;
			}
			
			return parseResult;
		}
		
		public static void Main (string[] args)
		{
			MagicParser mp = new MagicParser ();
			Console.WriteLine (mp.Parse(args[0], DateTime.Today));
		}
	}
}


src/Reminder.cs
//
// Reminder.cs: A tomboy plugin that displays a note at specified dates.
//
// Copyright 2005,  Raphaël Slinckx <raphael@slinckx.net>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
// http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

// $Id: Reminder.cs,v 1.4 2005/05/11 11:59:23 kikidonk Exp $

using System;
using System.Text.RegularExpressions;
using System.Collections;
using System.Runtime.InteropServices;

using Gtk;

using Tomboy;

using MagicString;

namespace Tomboy.Reminder
{
	public class Reminder : NoteAddin
	{
		//The reminder date expression, currently (in a non-langage specific fashion) it is
		// !date
		//or
		// remind date | alert: next monday | rappel mardi prochain
		//anywhere once in a line
		static string[] REMINDER_PATTERNS = {
			".*\\!",
			"(remind|rappel|alert)\\W+"
		};
		
		static Regex REMINDER =	new Regex ("^(("+
			REMINDER_PATTERNS[0]
			+")|("+
			REMINDER_PATTERNS[1]
			+"))(?<date>.*)$", RegexOptions.IgnoreCase | RegexOptions.Compiled | RegexOptions.Multiline);
		
		static MagicParser magicParser = new MagicParser();
		
		static Hashtable noteEvents = new Hashtable();
		
		static uint RE_REMIND_DELAY = 30*60; //Wait x before re-showing [in seconds] (snooze)
		
		public override void Initialize ()
		{
			// If a tag of this name already exists, don't install.
			if (Note.TagTable.Lookup ("reminder") == null) {
				TextTag highlightTag = new TextTag("reminder");
				highlightTag.Background = "yellow";
				
				Note.TagTable.Add (highlightTag);
			}
			
			ScanNote();
		}
				
		public override void Shutdown () {
			//Remove the delegate
			Window.UnmapEvent -= OnWindowClosed;
		}
		
		public override void OnNoteOpened () 
		{
			//Add a delegate when the user closes the window
			Window.UnmapEvent += OnWindowClosed;
			
			// Watch text changes to highlight valid patterns	
			Buffer.InsertText += OnInsertText;
			Buffer.DeleteRange += OnDeleteRange;

			ScanNote();
		}
			
		void OnWindowClosed (object sender, UnmapEventArgs args)
		{		
			//Parse the note to search for dates
			ScanNote();
		}

		void OnInsertText (object sender, Gtk.InsertTextArgs args)
		{
			SetupTimer(Note, args.Pos, args.Pos);
		}

		void OnDeleteRange (object sender, Gtk.DeleteRangeArgs args)
		{
			SetupTimer(Note, args.Start, args.End);
		}
		
		void SetupTimer(Note note, TextIter start, TextIter end)
		{
			if (!start.StartsLine())
				start.BackwardLine();
			if (!end.EndsLine())
				end.ForwardToLineEnd ();

			Buffer.RemoveTag("reminder", start, end);
			//Buffer.RemoveAllTags(start, end); // This breaks stuff - what purpose does it serve?
			SetupTimer(Note, start.GetSlice(end), start.Line);
		}
		
		protected void ScanNote()
		{
			//Delete every event associated with this note
			RemoveEvents(Note);
			
			string[] lines = Note.TextContent.Split('\n');
			int i = 0;
			foreach(string line in lines) {
				SetupTimer(Note, line, i++);
			}
		}
		
		private void RemoveEvents(Note note)
		{
			IList events = noteEvents[note] as IList;
			if(events == null)
				return;
				
			foreach(ReminderEvent ev in events) {
				ev.Cancel ();
			}
			noteEvents.Remove(note);
		}
		
		private void AddEvent (ReminderEvent ev) {
			//Retreive the events for this note
			IList events = noteEvents[ev.Note] as IList;
			if(events == null) {
				//No previous events for this note, create a new list
				events = new ArrayList();
				noteEvents[ev.Note] = events;
			}
			else if(events.Contains(ev))
				return;
			
			//Add the new events fo rthe note
			events.Add(ev);
			
			ev.Start();
		}
		
		private void HighlightEvent (Note note, int i)
		{
			NoteBuffer buf = note.Buffer;
			Console.WriteLine("Highlight line:"+i);
			Gtk.TextIter start = buf.GetIterAtLine(i);
			Gtk.TextIter end = start;
			end.ForwardToLineEnd();
			
			buf.ApplyTag ("reminder", start, end);
		}
		
		private void SetupTimer (Note note, string line, int i)
		{
			//Console.WriteLine("Reminder.cs:line:"+line);
			for (Match match = REMINDER.Match (line); match.Success; match = match.NextMatch ()) {
				ParseResult result;
				try {
					//Console.WriteLine("Reminder.cs:Parsing date:"+match.Groups ["date"].ToString());
					result = magicParser.Parse(match.Groups ["date"].ToString(), note.ChangeDate.Date);
				} catch(Exception e) {
					continue;
				}

				DateTime time = result.Date;	

				DateTime now = DateTime.Now;
				long due = (time.Ticks - now.Ticks)/10000;
				
				//Console.WriteLine("Reminder.cs:Got Reminder:"+now+" (due in: "+due+")");
				if(!result.HasTime && result.HasDate) {
					//The reminder was just a date, with no specified time, we remind the note
					//through the entire day
					if(time.Day == now.Day && time.Month == now.Month && time.Year == now.Year) {
						due = RE_REMIND_DELAY*1000; //Wait some time, before re-showing again
					}
				}
				
				//We have a past reminder, forget it.
				if(due < 0)
					continue;

				Console.WriteLine("Reminder.cs:Adding Reminder:"+time+" (due in: "+due/1000+" sec)");			
				AddEvent(new ReminderEvent(note, time, due));
				HighlightEvent(note, i);
			}
		}
		
		private class ReminderEvent {
			private Note note;
			private DateTime date;
			private long dueTime;
			private bool cancelled = false;
			
			public ReminderEvent(Note note, DateTime date, long dueTime) {
				this.note = note;
				this.date = date;
				this.dueTime = dueTime;
			}
			
			public void Start () {
				Console.WriteLine("Reminder.cs:Timer '"+note.Title+"' @ "+date);
				GLib.Timeout.Add((uint) dueTime, new GLib.TimeoutHandler(OnShowReminder));
			}
			
			public void Cancel () {
				cancelled = true;
			}
			
			[DllImport("libgtk-x11-2.0.so.0")]
			static extern void gtk_window_set_keep_above (IntPtr raw, bool setting);
			
			private bool OnShowReminder ()
			{
				if(! cancelled) {
					(noteEvents[note] as IList).Remove(this);
					Console.WriteLine("Reminder.cs:New Reminder: {0}", note.Title);
					
					Window w = note.Window;
					
					// Try to use the gtk lib to set the window on top
					try {
						gtk_window_set_keep_above (w.Handle, true);
					} catch (Exception e) {
						//Console.WriteLine ("Could not set on top: {0}", e);
					}
					
					w.Stick ();
					w.Present ();
				}
				
				return false;
			}
					
			public Note Note {
				get { return note; }
			}
			
			public DateTime Date {
				get { return date; }
			}
			
			public override bool Equals (object o) {
				ReminderEvent ev = o as ReminderEvent;
				return note.Equals(ev.Note) && date == ev.Date;
			}
			
			public override int GetHashCode () {
				return note.GetHashCode () ^ date.GetHashCode ();
			}
		}
	}
}